### PR TITLE
[Merged by Bors] - feat: Add `mixedEmbedding.norm`

### PR DESCRIPTION
**Pull request by xroblot**
Define a norm of the mixed space `ℝ^r₁ × ℂ^r₂` associated to a number field `K` of signature `(r₁, r₂)`. The norm is defined such that the norm of `mixedEmbedding K a` for `a : K` is equal to the absolute value of the norm of `a`
over `ℚ`. 

This norm will be used in a future PR about the quotient of the mixed space by the action of the units of `K`. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 1]
```diff
@@ -241,9 +241,73 @@ theorem disjoint_span_commMap_ker [NumberField K] :
...
+    simp_rw [Prod.fst_zero, Prod.snd_zero, Pi.zero_apply, norm_zero, zero_pow (two_ne_zero),
+      mul_eq_zero, Finset.prod_const, pow_eq_zero_iff', true_and, Finset.card_univ]
+    by_contra!
+    have : finrank ℚ K = 0 := by
+      rw [← card_add_two_mul_card_eq_rank, NrRealPlaces, NrComplexPlaces, this.1, this.2]
+    exact ne_of_gt finrank_pos this
+  map_mul' _ _ := by simp only [Prod.fst_mul, Pi.mul_apply, norm_mul, Real.norm_eq_abs,
+      Finset.prod_mul_distrib, Prod.snd_mul, Complex.norm_eq_abs, mul_pow]; ring
+
+protected theorem norm_ne_zero_iff {x : E K} :
```
Comment (faenuccio) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:269]: Why only a `ne_zero_iff` and not also some API like `norm_zero`, `norm_eq_zero_iff`and `norm_one`?

File: Mathlib/NumberTheory/NumberField/Embeddings.lean [hunk 2]
```diff
@@ -430,6 +430,10 @@ theorem embedding_of_isReal_apply {w : InfinitePlace K} (hw : IsReal w) (x : K)
     ((embedding_of_isReal hw) x : ℂ) = (embedding w) x :=
   ComplexEmbedding.IsReal.coe_embedding_apply (isReal_iff.mp hw) x
 
+theorem norm_embedding_eq_of_isReal {w : InfinitePlace K} (hw : IsReal w) (x : K) :
```
Comment (faenuccio) [inline Mathlib/NumberTheory/NumberField/Embeddings.lean:433]:
```suggestion
theorem norm_embedding_of_isReal {w : InfinitePlace K} (hw : IsReal w) (x : K) :
```

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean (hunk 1 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:269]: Well, I guess I only proved the results that I was going to need later on 😅  In any case, you are right and I should a couple more results.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 3]
```diff
@@ -241,9 +241,73 @@ theorem disjoint_span_commMap_ker [NumberField K] :
...
+    mixedEmbedding.norm x ≠ 0 ↔ (∀ w, x.1 w ≠ 0) ∧ (∀ w, x.2 w ≠ 0) := by
+  simp_rw [mixedEmbedding.norm, MonoidWithZeroHom.coe_mk, ZeroHom.coe_mk, mul_ne_zero_iff,
+    Finset.prod_ne_zero_iff, Finset.mem_univ, forall_true_left, pow_ne_zero_iff two_ne_zero,
+    norm_ne_zero_iff]
+
+theorem norm_const (c : ℝ) :
```
[the remark this answers is not in the document]
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:275]: Maybe scalar would be better? or just plain `real`?

[the remark this answers is not in the document]
Comment (faenuccio) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:275]: I think `real` would be best.

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean (hunk 1 above)
Comment (xroblot) [reply Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:280]: Here are some more lemmas.

**Comment by faenuccio** [comment]
Thanks, LGTM!

**Comment by xroblot** [comment]
> Thanks, LGTM!

Thanks for the review!

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 4]
```diff
@@ -241,9 +241,82 @@ theorem disjoint_span_commMap_ker [NumberField K] :
...
 
+noncomputable section norm
+
+open scoped Classical
+
+open BigOperators
+
+variable [NumberField K] {K}
+
+/-- The norm of `x` is `∏ w real, ‖x‖_w, ∏ w complex, ‖x‖_w ^ 2`. It is defined such that
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:252]:
```suggestion
/-- The norm of `x` is `∏ w real, ‖x‖_w * ∏ w complex, ‖x‖_w ^ 2`. It is defined such that
```

File: Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean [hunk 5]
```diff
@@ -241,9 +241,82 @@ theorem disjoint_span_commMap_ker [NumberField K] :
...
+    simp_rw [Prod.fst_zero, Prod.snd_zero, Pi.zero_apply, norm_zero, zero_pow (two_ne_zero),
+      mul_eq_zero, Finset.prod_const, pow_eq_zero_iff', true_and, Finset.card_univ]
+    by_contra!
+    have : finrank ℚ K = 0 := by
+      rw [← card_add_two_mul_card_eq_rank, NrRealPlaces, NrComplexPlaces, this.1, this.2]
+    exact ne_of_gt finrank_pos this
+  map_mul' _ _ := by simp only [Prod.fst_mul, Pi.mul_apply, norm_mul, Real.norm_eq_abs,
+      Finset.prod_mul_distrib, Prod.snd_mul, Complex.norm_eq_abs, mul_pow]; ring
+
+protected theorem norm_zero :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/CanonicalEmbedding/Basic.lean:269]: Do we really need these two results? `norm` is a morphism of groups with zero so everything should follow from the general machinery.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

**Pull request closed without merging**